### ISSUE TITLE
[BUG][mthreads] Miscompile equivalent kernel variant outputs inconsistent results

### ISSUE THREAD

**Issue by O1iveira7** [FlagTree/flagtree#982]
## Environment
  - Triton 3.6.0 (FlagTree mthreads3.6), MUSA backend, MTT S5000
  - torch_musa 2.7.1, MUSA 4.3.5, Python 3.10
## Symptom
  Semantically-identical transform (inject two runtime-dead branches — one
  masked store, one scalar store + `return` — plus a statement reorder and an
  always-true wrapper around the main store) makes the kernel output
  `[[-128, 127, -128], [-128, 127, -128]]` where the original kernel outputs
  the correct int8 quantization `[[-128, -128, -128], [-128, 127, 127]]` —
  2/6 elements flipped at the saturation boundary (`-128 <-> 127`).

  The transform is a semantic no-op (all injected code is runtime-dead: opaque
  predicates 0<1=true, 0>1=false, the dead branches never execute). The wrong
  output is silent — no error, no crash.

  Codegen-level heisenbug: adding a semantic no-op probe store (an extra
  store of the pre-clamp value) makes the mismatch disappear, and the probed
  pre-clamp values match the CPU fp64 reference bit-for-bit — the error is in
  code generation (register allocation / scheduling), not in kernel semantics.

**FlagTree mthreads3.6 output:**
>  original    = [[-128, **-128**, -128], [-128, 127, **127**]]
 > transformed = [[-128, **127**, -128], [-128, 127, **-128**]]

**Triton3.7 RTX3070 output:**
>original = [[-128, -128, -128], [-128, 127, 127]]
>reduced  = [[-128, -128, -128], [-128, 127, 127]]

  ## Minimal reproducer
```
● #!/usr/bin/env python3
  # -*- coding: utf-8 -*-
  """
  import sys

  import torch
  import triton
  import triton.language as tl

  M, N, K = 2, 3, 5
  # scale_A, scale_B, scale_C, zero_point_A, zero_point_B, zero_point_C
  SCALARS = (0.5, 0.25, 0.125, 1, -2, 3)
  A = [[-69, -84, 106, -42, -104], [-99, -30, 87, -28, 84]]
  B = [[112, -113, -62], [85, 113, 102], [-127, -26, -34], [-128, -81, -116], [111, 18, 123]]


  # ============================================================ original kernel
  @triton.jit
  def int8_quant_matmul_kernel(
      a_ptr,
      b_ptr,
      c_ptr,
      M,
      N,
      K,
      scale_A,
      scale_B,
      scale_C,
      zero_point_A,
      zero_point_B,
      zero_point_C,
      BLOCK_SIZE_M: tl.constexpr,
      BLOCK_SIZE_N: tl.constexpr,
      BLOCK_SIZE_K: tl.constexpr,
      GROUPSIZE: tl.constexpr,
  ):
      hw_pid0 = tl.program_id(0)
      hw_pid1 = tl.program_id(1)

      num_programs_pid0 = tl.num_programs(0)
      num_programs_pid1 = tl.num_programs(1)

      pid0, pid1 = tl.swizzle2d(
          hw_pid0, hw_pid1, num_programs_pid0, num_programs_pid1, GROUPSIZE
      )

      offset_M = pid0 * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
      offset_N = pid1 * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
      offset_K = tl.arange(0, BLOCK_SIZE_K)

      mask_M = offset_M < M
      mask_N = offset_N < N
      mask_K = offset_K < K

      offset_A = offset_M[:, None] * K + offset_K[None, :]
      mask_A = mask_M[:, None] & mask_K[None, :]

      offset_B = offset_K[:, None] * N + offset_N[None, :]
      mask_B = mask_K[:, None] & mask_N[None, :]

      accumulator = tl.zeros([BLOCK_SIZE_M, BLOCK_SIZE_N], dtype=tl.int32)
      accumulator_sum_A = tl.zeros([BLOCK_SIZE_M], dtype=tl.int32)
      accumulator_sum_B = tl.zeros([BLOCK_SIZE_N], dtype=tl.int32)

      scale_AB = scale_A * scale_B / scale_C

      for current_k_index in range(0, K, BLOCK_SIZE_K):
          if current_k_index + BLOCK_SIZE_K >= K:
              current_mask_K = (offset_K + current_k_index) < K
              mask_A = mask_M[:, None] & current_mask_K[None, :]
              mask_B = current_mask_K[:, None] & mask_N[None, :]

          data_A = tl.load(a_ptr + offset_A, mask=mask_A).to(tl.float32)
          data_B = tl.load(b_ptr + offset_B, mask=mask_B).to(tl.float32)

          accumulator += tl.dot(data_A, data_B).to(tl.int32)
          accumulator_sum_A += data_A.sum(1).to(tl.int32)
          accumulator_sum_B += data_B.sum(0).to(tl.int32)

          offset_A += BLOCK_SIZE_K
          offset_B += BLOCK_SIZE_K * N

      result = (
          accumulator
          - (accumulator_sum_A[:, None] * zero_point_B)
          - (accumulator_sum_B[None, :] * zero_point_A)
          + (K * zero_point_A * zero_point_B)
      )
      result = result.to(tl.float32) * scale_AB
      result = tl.floor(result + 0.5) + zero_point_C
      result = tl.clamp(result, -128, 127)

      offset_C = offset_M[:, None] * N + offset_N[None, :]
      mask_C = mask_M[:, None] & mask_N[None, :]

      tl.store(c_ptr + offset_C, result.to(tl.int8), mask=mask_C)


  def solve_original(
      a: torch.Tensor,
      b: torch.Tensor,
      c: torch.Tensor,
      M: int,
      N: int,
      K: int,
      scale_A: float,
      scale_B: float,
      scale_C: float,
      zero_point_A: int,
      zero_point_B: int,
      zero_point_C: int,
  ):
      BLOCK_SIZE_M = 128
      BLOCK_SIZE_N = 128
      BLOCK_SIZE_K = 64
      GROUP_SIZE = 8

      grid = (triton.cdiv(M, BLOCK_SIZE_M), triton.cdiv(N, BLOCK_SIZE_N))
      int8_quant_matmul_kernel[grid](
          a,
          b,
          c,
          M,
          N,
          K,
          scale_A,
          scale_B,
          scale_C,
          zero_point_A,
          zero_point_B,
          zero_point_C,
          BLOCK_SIZE_M,
          BLOCK_SIZE_N,
          BLOCK_SIZE_K,
          GROUP_SIZE,
      )


  # ============================================ reduced variant (min_final shape)
  # delta-reduce minimal trigger: two runtime-false dead branches (masked
  # store / scalar store + return) + main-path statement permute + main
  # store wrapped in an always-true branch. All other injections removed.
  @triton.jit
  def int8_quant_matmul_kernel_min(_fz_switch_ptr, _fz_switch_output_ptr, a_ptr, b_ptr, c_ptr,
                                   M, N, K, scale_A, scale_B, scale_C, zero_point_A, zero_point_B,
                                   zero_point_C, BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr,
                                   BLOCK_SIZE_K: tl.constexpr, GROUPSIZE: tl.constexpr):
      opaque_values = tl.load(_fz_switch_ptr + tl.arange(0, 2))
      (opaque_zero, opaque_one) = tl.split(opaque_values)
      opaque_true = opaque_zero < opaque_one
      opaque_false = opaque_zero > opaque_one
      hw_pid0 = tl.program_id(0)
      hw_pid1 = tl.program_id(1)
      num_programs_pid0 = tl.num_programs(0)
      num_programs_pid1 = tl.num_programs(1)
      (pid0, pid1) = tl.swizzle2d(hw_pid0, hw_pid1, num_programs_pid0, num_programs_pid1, GROUPSIZE)
      offset_M = pid0 * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)
      offset_N = pid1 * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)
      # ---- dead branch 1: runtime-false (opaque_false = 0 > 1), masked store ----
      if opaque_false:
          tl.store(_fz_switch_output_ptr + 34 + tl.arange(0, 2), opaque_values.to(tl.float32),
                   mask=(tl.program_id(axis=0) == 0) & (tl.program_id(axis=1) == 0) & (tl.program_id(axis=2) == 0))
      offset_K = tl.arange(0, BLOCK_SIZE_K)
      # ---- dead branch 2: runtime-false, scalar store + return ----
      if opaque_false:
          _fz_tiny = opaque_zero + opaque_one
          tl.store(_fz_switch_output_ptr + 10 + tl.arange(0, 2), _fz_tiny.to(tl.float32),
                   mask=(tl.program_id(axis=0) == 0) & (tl.program_id(axis=1) == 0) & (tl.program_id(axis=2) == 0))
          return
      mask_M = offset_M < M
      mask_N = offset_N < N
      mask_K = offset_K < K
      # ---- permute: statement reordering (semantically equivalent to original) ----
      offset_A = offset_M[:, None] * K + offset_K[None, :]
      mask_B = mask_K[:, None] & mask_N[None, :]
      mask_A = mask_M[:, None] & mask_K[None, :]
      offset_B = offset_K[:, None] * N + offset_N[None, :]
      accumulator = tl.zeros([BLOCK_SIZE_M, BLOCK_SIZE_N], dtype=tl.int32)
      accumulator_sum_A = tl.zeros([BLOCK_SIZE_M], dtype=tl.int32)
      accumulator_sum_B = tl.zeros([BLOCK_SIZE_N], dtype=tl.int32)
      scale_AB = scale_A * scale_B / scale_C
      for current_k_index in range(0, K, BLOCK_SIZE_K):
          if current_k_index + BLOCK_SIZE_K >= K:
              current_mask_K = offset_K + current_k_index < K
              mask_A = mask_M[:, None] & current_mask_K[None, :]
              mask_B = current_mask_K[:, None] & mask_N[None, :]
          data_A = tl.load(a_ptr + offset_A, mask=mask_A).to(tl.float32)
          data_B = tl.load(b_ptr + offset_B, mask=mask_B).to(tl.float32)
          accumulator += tl.dot(data_A, data_B).to(tl.int32)
          accumulator_sum_A += data_A.sum(1).to(tl.int32)
          accumulator_sum_B += data_B.sum(0).to(tl.int32)
          offset_A += BLOCK_SIZE_K
          offset_B += BLOCK_SIZE_K * N
      result = (
          accumulator
          - (accumulator_sum_A[:, None] * zero_point_B)
          - (accumulator_sum_B[None, :] * zero_point_A)
          + (K * zero_point_A * zero_point_B)
      )
      result = result.to(tl.float32) * scale_AB
      result = tl.floor(result + 0.5) + zero_point_C
      result = tl.clamp(result, -128, 127)
      offset_C = offset_M[:, None] * N + offset_N[None, :]
      mask_C = mask_M[:, None] & mask_N[None, :]
      # ---- store wrap: always-true runtime branch ----
      _fz_a4_loop_once = 0
      if opaque_true:
          tl.store(c_ptr + offset_C, result.to(tl.int8), mask=mask_C)
      else:
          pass


  def solve_reduced(
      a: torch.Tensor,
      b: torch.Tensor,
      c: torch.Tensor,
      M: int,
      N: int,
      K: int,
      scale_A: float,
      scale_B: float,
      scale_C: float,
      zero_point_A: int,
      zero_point_B: int,
      zero_point_C: int,
  ):
      _fz_device = next((_fz_value.device for _fz_value in (a, b, c, M, N, K, scale_A, scale_B,
                         scale_C, zero_point_A, zero_point_B, zero_point_C)
                         if isinstance(_fz_value, torch.Tensor)), None)
      if _fz_device is None:
          raise RuntimeError('no tensor argument in solve() to select the opaque-buffer device')
      _fz_switch_ptr = torch.tensor([0, 1], dtype=torch.int32, device=_fz_device)
      _fz_switch_output_ptr = torch.empty(400, dtype=torch.float32, device=_fz_device)
      BLOCK_SIZE_M = 128
      BLOCK_SIZE_N = 128
      BLOCK_SIZE_K = 64
      GROUP_SIZE = 8
      grid = (triton.cdiv(M, BLOCK_SIZE_M), triton.cdiv(N, BLOCK_SIZE_N))
      int8_quant_matmul_kernel_min[grid](_fz_switch_ptr, _fz_switch_output_ptr, a, b, c, M, N, K,
                                         scale_A, scale_B, scale_C, zero_point_A, zero_point_B,
                                         zero_point_C, BLOCK_SIZE_M, BLOCK_SIZE_N, BLOCK_SIZE_K,
                                         GROUP_SIZE)


  def main() -> int:
      if not torch.musa.is_available():
          print("MUSA is unavailable", file=sys.stderr)
          return 2

      a0 = torch.tensor(A, dtype=torch.int8, device="musa")
      b0 = torch.tensor(B, dtype=torch.int8, device="musa")
      try:
          original = torch.zeros((M, N), dtype=torch.int8, device="musa")
          reduced = torch.zeros((M, N), dtype=torch.int8, device="musa")
          solve_original(a0, b0, original, M, N, K, *SCALARS)
          solve_reduced(a0, b0, reduced, M, N, K, *SCALARS)
          torch.musa.synchronize()
      except Exception as e:
          msg = str(e)
          print(f"FAIL: {type(e).__name__}: {msg.splitlines()[0][:80]}")
          return 1

      original = original.cpu().tolist()
      reduced = reduced.cpu().tolist()
      print(f"original = {original}")
      print(f"reduced  = {reduced}")
      return 0 if original != reduced else 1


  if __name__ == "__main__":
      raise SystemExit(main())
```
  ### Difference between the two kernels
Here are the main differences between the original kernel int8_quant_matmul_kernel and the reduced variant int8_quant_matmul_kernel_min:

1. Extra opaque-switch parameters. The variant takes two extra pointers — _fz_switch_ptr (a device tensor [0, 1]) and _fz_switch_output_ptr (a scratch buffer). It loads opaque_values, splits them, and derives two runtime-determined booleans that the compiler cannot fold at compile time: opaque_true (0 < 1, always True at runtime) and opaque_false (0 > 1, always False at runtime).

2. Two runtime-dead branches. Two if opaque_false: blocks are injected near the top. They never execute, but they force the MTGPU codegen to emit live branches:
- branch 1: a masked store into the scratch buffer,
- branch 2: a scalar store plus a return (dead-jmp).

3. Main-path statement permute. The computation itself is untouched, but statements are reordered — e.g. mask_B/mask_A are computed before offset_B, the accumulator/scale setup is hoisted before the for loop. Semantically identical to the original.

4. Main store wrapped in an always-true branch. The single real tl.store(c_ptr + offset_C, ...) is moved inside if opaque_true: ... else: pass instead of being emitted unconditionally.

In short: no arithmetic or control-flow in the main path changes. The variant adds only an opaque runtime switch plus never-executed/always-executed wrapper branches around a semantically-identical main body. On MUSA, that shape triggers the -128 ↔ 127 saturation-flip miscompile; on CUDA the same source compiles and runs correctly.

 ### evidence (NOT CONFIRED GENERATED BY CLAUDE CODE)
● #### Root-cause assessment

  **MUSA backend codegen defect (register allocation / instruction scheduling),
  not a semantic error.**

  Key evidence:

  1. **Probe heisenbug (decisive)**: adding a semantic no-op probe store (an
     extra store of the pre-clamp value) to the *same* source makes the output
     match the CPU fp64 reference bit-for-bit, and the probed pre-clamp values
     are also bit-exact. Same semantics, different codegen → different result.
     The error lives in the codegen path, not in kernel semantics.
  2. **All injections are semantic no-ops**: the dead branches (runtime-false
     `0 > 1`) never execute; permute / store-wrap are equivalent; the TTGIR
     main path is structurally identical to the original.
  3. **Error magnitude is thousands**: (0,1) needs +3348 to flip to 127,
     (1,2) needs −13836 to flip to −128 — not floating-point rounding; with
     random inputs the wrong output always satisfies `row0 == row1`, i.e.
     some lanes receive row-shifted data (dot/sum intermediates crossed).
  4. **Non-monotonic trigger**: 63/64 of a 64-combination subset sweep still
     triggered — the trigger is a global function of the code shape, not a
     defect in a single op.
  5. **mubin grows 21%** (320KB → 388KB): codegen output is significantly
     perturbed by the injected dead code.

  #### Reasoning

  saturation-boundary flips (2/6, -128 <-> 127)
    └─ pre-clamp values off by thousands → data misrouting, not rounding
        └─ semantics ruled out: all injections no-op, main path identical
        └─ probe: semantics unchanged, codegen changed → result becomes correct
            └─ error determined by codegen outcome → register alloc / scheduling
                └─ dead branches perturb allocation → some lanes' accumulators corrupted
                    └─ masked loads without other= amplify (OOB lane values drift)

  #### Difference from musa-issue-000004

  - 000004: **stable structural trigger** — `cf.cond_br` + `llvm.assume` in
    TTGIR→LLIR makes the sort network's compare/selects deleted wholesale
    (LLIR shrinks).
  - 000005: **perturbation-sensitive** — a change in scheduling flips the
    outcome; points to register allocation / scheduling, not instruction
    elimination (mubin grows instead of shrinking).
```

**Comment by zhzhcookie** [comment]
Please try `flagtree===0.6.2a3+mthreads3.6`, or build flagtree from source.
